host/clearkey_platform.c

```diff
@@ -0,0 +1,162 @@
+ */
+#ifdef USE_SEDGET_VIDEO
+#include "sedget_video.h"
+#elif defined(SDP_USES_NATIVE_HANDLE)
+#include <cutils/native_handle.h>
+#endif
+
+int clearkey_plat_get_mem_fd(void *mem_handle)
+{
+#ifdef USE_SEDGET_VIDEO
+	return sedget_get_mem_fd((sedget_protected_buffer *)mem_handle);
+#elif defined(SDP_USES_NATIVE_HANDLE)
+	native_handle_t *handle = (native_handle_t *)mem_handle;
+	return handle->data[0];
+#else
+	return -1;
+#endif
+}
+
+#ifdef SDP_PROTOTYPE
+
+/* SDP protoype test functions to help validate
+ * decryption into a SDP buffer on platforms which
+ * don't already have a secure codec and therefore
+ * don't get passed a buffer to decrypt method */
+
+int allocate_ion_buffer(size_t size, int heap_id)
+{
+	struct ion_allocation_data alloc_data;
+	struct ion_handle_data hdl_data;
+	struct ion_fd_data fd_data;
+	int ion;
+	int fd = -1;
+
+	ion = open("/dev/ion", O_RDWR);
+	if (ion < 0) {
+		FP("Error; failed to open /dev/ion\n");
+		return fd;
+	}
+
+	if (heap_id < 0)
+	  heap_id = ION_HEAP_TYPE_UNMAPPED;
+
+	FP("Allocate in ION heap '%s'\n",
+		heap_id == ION_HEAP_TYPE_SYSTEM ? "system" :
+		heap_id == ION_HEAP_TYPE_SYSTEM_CONTIG ? "system contig" :
+		heap_id == ION_HEAP_TYPE_CARVEOUT ? "carveout" :
+		heap_id == ION_HEAP_TYPE_CHUNK ? "chunk" :
+		heap_id == ION_HEAP_TYPE_DMA ? "dma" :
+		heap_id == ION_HEAP_TYPE_UNMAPPED ? "unmapped" :
+		"custom");
+
+	alloc_data.len = size;
+	alloc_data.align = 0;
+	alloc_data.flags = 0;
+	alloc_data.heap_id_mask = 1 << heap_id;
+	if (ioctl(ion, ION_IOC_ALLOC, &alloc_data) == -1)
+		goto out;
+
+	fd_data.handle = alloc_data.handle;
+	if (ioctl(ion, ION_IOC_SHARE, &fd_data) != -1)
+		fd = fd_data.fd;
+
+	hdl_data.handle = alloc_data.handle;
+	(void)ioctl(ion, ION_IOC_FREE, &hdl_data);
+out:
+	close(ion);
+	return fd;
+}
+
+int ion_map_and_memcpy(unsigned char* out_data, uint32_t length, int secure_fd)
+{
+  /* To allow prototyping SDP, map the secure buffer
+     (it is not actually protected) and copy data into the destination
+     buffer. */
+  ion_user_handle_t ion_user_handle = -1;
+  unsigned char *mapped_secure_buf = NULL;
+  int map_fd = -1;
+  int ion_fd;
+  int ret = 0;
+
+  ion_fd = open("/dev/ion", O_RDWR);
+  if (ion_fd < 0) {
+    FP("Error; failed to open /dev/ion\n");
+    ret = ion_fd;
+    goto error;
+  }
+
+  ret = ion_import(ion_fd, secure_fd, &ion_user_handle);
+  if (ret < 0) {
+    FP("Cannot import secure buffer\n");
+    close(ion_fd);
+    goto error;
+  }
+
+  ret = ion_map(ion_fd, ion_user_handle, length, PROT_READ | PROT_WRITE,
+		MAP_SHARED, 0, &mapped_secure_buf, &map_fd);
+  if(ret < 0) {
+    FP("Cannot map secure buffer\n");
+    ion_free(ion_fd, ion_user_handle);
+    ion_user_handle = -1;
+    close(ion_fd);
+    goto error;
+  }
+
+  memcpy(out_data, mapped_secure_buf, length);
+
+  ion_free(ion_fd, ion_user_handle);
+  ion_user_handle = -1;
+  munmap(mapped_secure_buf, length);
+  close(map_fd);
+  close(ion_fd);
+
```

Comment:
Nitpick.
Could the above be put like this (doing munmap() first, and ion_free() next):

```
  ret = ion_import(ion_fd, secure_fd, &ion_user_handle);
  if (ret < 0) {
    FP("Cannot import secure buffer\n");
    close(ion_fd);
    goto error1;
  }

  ret = ion_map(ion_fd, ion_user_handle, length, PROT_READ | PROT_WRITE,
		MAP_SHARED, 0, &mapped_secure_buf, &map_fd);
  if(ret < 0) {
    FP("Cannot map secure buffer\n");
    goto error2;
  }

  memcpy(out_data, mapped_secure_buf, length);

  munmap(mapped_secure_buf, length);
error2:
  ion_free(ion_fd, ion_user_handle);
  ion_user_handle = -1;
  close(map_fd);
error1:
  close(ion_fd);￼ 
```
If this doesn't make much sense, this part cam be left as it is now - it's up to you.

Reply:
Will fix